src/Microsoft.AspNet.Mvc.Razor/Buffer/RazorBuffer.cs

```diff
@@ -0,0 +1,158 @@
+    public class RazorBuffer : IHtmlContentBuilder
+    {
+        private readonly IRazorBufferScope _bufferScope;
+        private readonly string _name;
+        private int _currentIndex;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="RazorBuffer"/>.
+        /// </summary>
+        /// <param name="bufferScope">The <see cref="IRazorBufferScope"/>.</param>
+        /// <param name="name">A name to identify this instance.</param>
+        public RazorBuffer(IRazorBufferScope bufferScope, string name)
+        {
+            if (bufferScope == null)
+            {
+                throw new ArgumentNullException(nameof(bufferScope));
+            }
+
+            _bufferScope = bufferScope;
+            _name = name;
+        }
+
+        /// <summary>
+        /// Gets the backing buffer.
+        /// </summary>
+        public IList<RazorBufferSegment> BufferChunks { get; private set; }
+
+        /// <inheritdoc />
+        public IHtmlContentBuilder Append(string unencoded)
+        {
+            if (unencoded == null)
+            {
+                return this;
+            }
+
+            EnsureCapacity();
+
+            var chunk = BufferChunks[BufferChunks.Count - 1];
+            chunk.Data.Array[chunk.Data.Offset + _currentIndex] = new RazorValue(unencoded);
+            _currentIndex = (_currentIndex + 1) % chunk.Data.Count;
+            return this;
+        }
+
+        /// <inheritdoc />
+        public IHtmlContentBuilder Append(IHtmlContent content)
+        {
+            if (content == null)
+            {
+                return this;
+            }
+
+            EnsureCapacity();
+
+            var chunk = BufferChunks[BufferChunks.Count - 1];
+            chunk.Data.Array[chunk.Data.Offset + _currentIndex] = new RazorValue(content);
+            _currentIndex = (_currentIndex + 1) % chunk.Data.Count;
+            return this;
+        }
+
+        /// <inheritdoc />
+        public IHtmlContentBuilder AppendHtml(string encoded)
+        {
+            if (encoded == null)
+            {
+                return this;
+            }
+
+            EnsureCapacity();
+
+            var chunk = BufferChunks[BufferChunks.Count - 1];
+            chunk.Data.Array[chunk.Data.Offset + _currentIndex] = new RazorValue(new HtmlString(encoded));
+            _currentIndex = (_currentIndex + 1) % chunk.Data.Count;
+            return this;
+        }
+
+        /// <inheritdoc />
+        public IHtmlContentBuilder Clear()
+        {
+            if (BufferChunks != null)
+            {
+                _currentIndex = 0;
+                BufferChunks.Clear();
+            }
+
+            return this;
+        }
+
+        /// <inheritdoc />
+        public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+        {
+            if (BufferChunks == null)
+            {
+                return;
+            }
+
+            var htmlTextWriter = writer as HtmlTextWriter;
```

Comment:
Making the suggested change exposed a potential issue. We end up with an instance of `RazorBuffer` inside of a `CacheTagHelper` and we cannot keep the `RazorBuffer` around between requests.